### ISSUE TITLE
Add `regexp/prefer-result-array-groups` rule

### ISSUE THREAD

**Issue by RunDevelopment** [ota-meshi/eslint-plugin-regexp#347]
From [this comment](https://github.com/ota-meshi/eslint-plugin-regexp/pull/339#discussion_r719249323).

> `prefer-regexp-result-array-groups` is a rule that inherits some named capturing group checks from `no-unused-capturing-group`. 
> This rule prefers named access over array index access when using named capturing group in regex. Perhaps the process is similar to `prefer-named-replacement`.
> When adding `prefer-regexp-result-array-groups`, exclude the same check from `no-unused-capturing-group`.
> 
> The example code that `prefer-regexp-result-array-groups` warns is:
> 
> ```js
> const reFoo = /(?<foo>foo)/
> let resArr
> while (resArr = reFoo.exec(str)) {
>   const foo = resArr[1]
>   //                ^^^ Use `resArr.groups.foo` instead.
>   //                Currently it is warned by no-unused-capturing-group, but change it to not warn.
> }
> ```
> 
> The sample code that the new rule does not warn is:
> 
> ```js
> const reFoo = /(?<foo>foo)/
> let resArr
> while (resArr = reFoo.exec(str)) {
>   const foo = resArr.groups.foo // correct
> }
> ```
> 
> ```js
> const reFoo = /(foo)/
> let resArr
> while (resArr = reFoo.exec(str)) {
>   const foo = resArr[1] // Not using named capturing group in regex.
> }
> ```
> 
> ```js
> const reFoo = /(foo)/
> let resArr
> while (resArr = reFoo.exec(str)) {
>   const match = resArr[0] // The capturing group was not used. It's warned by no-unused-capturing-group rule.
> }
> ```
> 
> The `prefer-regexp-result-array-groups` rule checks not only the `exec` but also the return values of `match` and `matchAll`.

**Comment by RunDevelopment** [comment]
I think we should change the name a little. This whole plugin is about regexes, so I think `prefer-result-array-groups` is enough.

**Comment by ota-meshi** [comment]
`prefer-result-array-groups` is good :+1:! `regexp` is in the prefix.
